### PR TITLE
[JUJU-856] Merge 2.9 into develop

### PR DESCRIPTION
**Pull request by SimonRichardson**
This was a tricky merge, as we had a LOT of changes again, even after doing a merge forward yesterday.

756ae5bbf8 (upstream/2.9, origin/2.9, 2.9) Merge pull request #13918 from barrettj12/create-subnet2
4d4b89ad05 Merge pull request #13917 from barrettj12/no-stable-release
e5355c3297 Merge pull request #13909 from barrettj12/model-config-key-2.9
6490a1c47f Merge pull request #13912 from ycliuhw/enhance-setup.iss
73f833af91 Merge pull request #13913 from jujubot/increment-to-2.9.29
1f6af557dd Merge pull request #13906 from arnodel/juju-842-fix-mockgen
cb8d74d02c Merge pull request #13877 from jack-w-shaw/JUJU-815_Resolve_noop_transaction
e1ee28cc50 Merge pull request #13911 from wallyworld/replicaset-update-address
39ffa181b9 Merge pull request #13910 from juju/revert-13903-strip-buildnumber4agentversion
422670a4f5 Merge pull request #13908 from hmlanigan/watch-machines-only
385df47be4 Merge pull request #13867 from SimonRichardson/remove-facade-reflection
d4d38ca64e Merge pull request #13904 from tlm/docker-makefile-oneshot
15211adcd9 Merge pull request #13903 from ycliuhw/strip-buildnumber4agentversion
52ed3156d8 Merge pull request #13895 from manadart/2.9-netplan-skip-ipv6-link-local-ns
5a937f3b86 Merge pull request #13894 from tlm/docker-makefile-oneshot
014bb457e5 Merge pull request #13899 from hmlanigan/update-deploy-tests
fbfc86b074 Merge pull request #13902 from barrettj12/deploy-rev-ci
ddb78ad037 Merge pull request #13901 from barrettj12/refresh-idempotent
4cc2c21d1a Merge pull request #13900 from hmlanigan/fix-the-fix
d0fa33fa6a Merge pull request #13887 from hmlanigan/cmd-log-label

CONFLICT (content): Merge conflict in version/version.go
CONFLICT (content): Merge conflict in snap/snapcraft.yaml
CONFLICT (content): Merge conflict in scripts/win-installer/setup.iss
CONFLICT (content): Merge conflict in apiserver/facades/controller/statushistory/api.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/remoterelations/remoterelations.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/migrationmaster/facade.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/instancepoller/instancepoller.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/firewaller/firewaller.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/crossmodelrelations/crossmodelrelations.go
CONFLICT (content): Merge conflict in apiserver/facades/controller/actionpruner/pruner.go
CONFLICT (content): Merge conflict in apiserver/facades/client/subnets/subnets.go
CONFLICT (content): Merge conflict in apiserver/facades/client/storage/storage.go
CONFLICT (content): Merge conflict in apiserver/facades/client/sshclient/facade.go
CONFLICT (content): Merge conflict in apiserver/facades/client/spaces/spaces.go
CONFLICT (content): Merge conflict in apiserver/facades/client/resources/facade.go
CONFLICT (content): Merge conflict in apiserver/facades/client/modelmanager/modelmanager.go
CONFLICT (content): Merge conflict in apiserver/facades/client/modelgeneration/modelgeneration.go
CONFLICT (content): Merge conflict in apiserver/facades/client/modelconfig/modelconfig.go
CONFLICT (content): Merge conflict in apiserver/facades/client/machinemanager/machinemanager.go
CONFLICT (content): Merge conflict in apiserver/facades/client/controller/export_test.go
CONFLICT (content): Merge conflict in apiserver/facades/client/controller/controller.go
CONFLICT (content): Merge conflict in apiserver/facades/client/cloud/cloud.go
CONFLICT (content): Merge conflict in apiserver/facades/client/charms/client.go
CONFLICT (modify/delete): apiserver/facades/client/charmhub/charmhub.go deleted in HEAD and modified in 2.9. Version 2.9 of apiserver/facades/client/charmhub/charmhub.go left in tree.
CONFLICT (content): Merge conflict in apiserver/facades/client/bundle/bundle.go
CONFLICT (content): Merge conflict in apiserver/facades/client/backups/shim.go
CONFLICT (content): Merge conflict in apiserver/facades/client/applicationoffers/applicationoffers.go
CONFLICT (content): Merge conflict in apiserver/facades/client/application/application.go
CONFLICT (content): Merge conflict in apiserver/facades/client/action/action.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/upgradesteps/upgradesteps.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/upgradeseries/upgradeseries.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/uniter/uniter.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/storageprovisioner/shim.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/proxyupdater/proxyupdater.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/provisioner/provisioner.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/meterstatus/meterstatus.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/machine/machiner.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/instancemutater/instancemutater.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/credentialvalidator/credentialvalidator.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/caasoperator/operator.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/caasagent/caasagent.go
CONFLICT (content): Merge conflict in apiserver/facades/agent/agent/agent.go
CONFLICT (content): Merge conflict in apiserver/common/modelstatus_test.go
CONFLICT (content): Merge conflict in apiserver/allfacades.go

## QA steps

Tests Pass

### CODE REVIEW

**Review by manadart — COMMENTED** [review]
Too many files to review effectively, but it *looks* OK except for the registration issue, which is causing a test failure.

File: apiserver/facades/agent/agent/register.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// Register is called to expose a package of facades onto a given registry.
+func Register(registry facade.FacadeRegistry) {
+	registry.MustRegister("Agent", 3, func(ctx facade.Context) (facade.Facade, error) {
+		return newAgentAPIV3(ctx)
+	}, reflect.TypeOf((*AgentAPI)(nil)))
+}
+
+// newAgentAPIV3 returns an object implementing version 2 of the Agent API
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/agent/agent/register.go:24]: I'll fix this in a driveby later.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**